### DISCUSSION TITLE
Exercise 24: Build a run of the Ben-Or randomised consensus algorithm that never terminates.

### DISCUSSION THREAD

**Discussion by rotiroti** [sapienzastudentsnetwork/distributed-systems#49] [Exercises]
Build a run of the Ben-Or randomised consensus algorithm that never terminates.

**Comment by simonesestito** [comment]
Can we assume a faulty random number generator?

**Comment by susannacifani** [comment]
I found an example of a run of the Ben-Or that never terminates online:

![image](https://github.com/sapienzastudentsnetwork/distributed-system-2023-2024/assets/73530772/95615b7b-3b42-4efb-9ceb-d1e31d8e3a29)

**Comment by aleiis** [comment]
The Ben-Or algorithm is designed to converge with a high probability within a finite number of rounds. The only way to prolong the state of indecision is when no majority is reached, resulting in no ratified value. The presence of more failures makes it increasingly challenging to achieve a majority consensus. However, a majority can still be reached if $N-F$ participants share the same preference. In such a scenario, where $F$ equals the floor of $\frac{(N-1)}{2}$, the random number generator should never produce the same number as new preferences for $N-F$ participants, causing the algorithm to continue indefinitely.

Obviously, the algorithm wouldn't make progress if $F > \frac{N}{2}$, so it also wouldn't ever terminate.

**Comment by cornflexxx** [comment]
Consider the following setting: an even number of processes, half of them choose $0$ and the other half $1$. In the first phase we can't reach a majority so all of them send the message ($2$, $round$, $?$). In the second phase none of them outputs or ratify, so they all flip a the coin. Consider the following event: 

$$E = \text{all processes flip a coin, but neither 0 nor 1 achieves a majority} $$

clearly $\Pr[E]>0$.
Hence, there exists a run starting with such a setting and proceed indefinitely through a sequence of events $E$. With $round \to \infty$ the probability of  this  non-terminating execution tends to $0$.
